qiskit_dynamics/models/operator_collections.py

```diff
@@ -0,0 +1,287 @@
+    @abstractmethod
+    def hilbert_space_dimension(self) -> int:
+        """Gets dimension of hilbert space of system"""
+        pass
+
+    @abstractmethod
+    def evaluate_without_state(self, signal_values: Array) -> Array:
+        r"""If the model can be represented simply and
+        without reference to the state involved, e.g.
+        in the case \dot{y} = G(t)y(t) being represented
+        as G(t), returns this independent representation.
+        If the model cannot be represented in such a
+        manner (c.f. Lindblad model), then errors."""
+        pass
+
+    @abstractmethod
+    def evaluate_with_state(self, signal_values: Union[List[Array], Array], y: Array) -> Array:
+        """Evaluates the model for a given state
+        y provided the values of each signal
+        component s_j(t). Must be defined for all
+        models."""
+        pass
+
+    @abstractmethod
+    def apply_function_to_operators(self, function_to_apply: Optional[Callable]) -> None:
+        """Apply some function to all operators.
+        Useful for e.g. basis transformations.
+
+        Args:
+            function_to_apply: Callable function
+            to be applied to each operator. Should
+            take in Array and return Array."""
+        pass
+
+    def __call__(
+        self, signal_values: Union[List[Array], Array], y: Optional[Array] = None
+    ) -> Array:
+        """Evaluates the model given the values of the signal
+        terms s_j, suppressing the choice between
+        evaluate_with_state and evaluate_without_state
+        from the user. May error if y is not provided and
+        model cannot be expressed without choice of state.
+        """
+
+        if y is None:
+            return self.evaluate_without_state(signal_values)
+
+        return self.evaluate_with_state(signal_values, y)
+
+    def copy(self):
+        """Return a copy of self."""
+        return deepcopy(self)
+
+
+class DenseOperatorCollection(BaseOperatorCollection):
+    r"""Meant to be a calculation object for models that
+    only ever need left multiplication–those of the form
+    \dot{y} = G(t)y(t), where G(t) = \sum_j s_j(t) G_j + G_d.
+    Can evaluate G(t) independently of y.
+    """
+
+    @property
+    def operators(self) -> Array:
+        return self._operators
+
+    @property
+    def num_operators(self) -> int:
+        return self._operators.shape[-3]
+
+    @property
+    def hilbert_space_dimension(self) -> int:
+        return self._operators.shape[-2]
+
+    def apply_function_to_operators(self, function_to_apply: Optional[Callable]):
+        """Applies a function to all operators in the collection.
+        Args:
+            function_to_apply: the function to be applied to all operators"""
+        self._operators = function_to_apply(self._operators)
+        if self.drift is not None:
+            self.drift = function_to_apply(self.drift)
+
+    def evaluate_without_state(self, signal_values: Array) -> Array:
```

Comment:
Similar to the comment for `GeneratorModel`, I'd rather call this somethign like: `evaluate_linear_combination`, and the `evaluate_with_state` to something like `apply_map`. Just to be more concrete about what's being "evaluated", and to avoid naming collisions for meanings of the word "state".